text/0000-location.md

```diff
@@ -0,0 +1,409 @@
+
+The primary object stored in state is �Location�, which consists of a 
+location_id (GS1 GLN), an owner (org_id compatible with Pike), and a list of 
+property name/value pairs. The properties available are defined by the Grid 
+Location Schema Transaction Family and are restricted to the fields and rules of
+the GS1 Location Property Schema. Transactions which are responsible for setting
+location state values must ensure that the properties conform with the 
+requirements of the GS1 Location Property Schema.
+
+The GS1 GLN is an identifier (location_id) used to identify entities or 
+locations. A GLN is the data transmitted from a data carrier (barcode, rfid, 
+etc.) scan and is made up of a company prefix, location reference, and check 
+digit.
+
+### Referencing Locations
+
+Locations are uniquely referenced by their location_id and location_namespace.
+For GS1, locations are referenced by the GLN identifier.
+
+**Note: For the remainder of this document, �address� will refer to a 
+Grid/Transact address as opposed to a Street Address.**
+
+### Location Addressing in the Merkle-Radix State System
+
+In order to uniquely locate GS1 locations in the Merkle-Radix state system, an 
+address must be constructed which identifies the storage location of the 
+Location representation.
+
+All Grid addresses are prefixed by the 6-hex-character namespace prefix �621dee�.
+Locations are further prefixed under the Grid namespace with reserved 
+enumerations of �04� (�00�, �01�, �02� and �03� being reserved for other 
+purposes) indicating �Locations� and an additional �01� indicating �GS1 
+Locations�.
+
+Therefore, all addresses starting with �621dee� + �04� are Grid locations, and 
+more specifically, all addresses starting with �621dee� + �04� + �01� are Grid 
+GS1 Locations identified by a GLN.
+
+The GLN format consists of a 13-digit �numeric string� that follows a similar 
+structure to GTIN-13. The GLN structure first includes a GS1 Company Prefix 
+which is 7-10 digits in length and is assigned by a GS1 Member Organization. A 
+Location Reference number follows which is 2-5 digits in length and is 
+allocated by the company to a location or party. Lastly, a single digit Check 
+Digit is calculated and applied according to a GS1 algorithm.
+
+![GLN Data Format](https://encrypted-tbn0.gstatic.com/images?q=tbn%3AANd9GcSFo3J5QULgkCWyIH3yeddRPx1H5aQusLRRq47y1n2FqTshGxs%3Ax-raw-image%3A%2F%2F%2Ff4fc819afc288cee1948ca0d40d76c3bcd56a6d270fcfeeac80fd25784eb6071&usqp=CAU)
+
+After the 10-hex-characters that are consumed by the grid namespace prefix, the 
+location, and GS1 prefixes, there are 60 hex characters remaining in the address.
+The 13 digits of the GLN can be left padded with 45-hex-character zeroes and 
+right padded with 2-hex-character zeroes to accommodate potential future storage
+associated with the GS1 Location representation, for example:
+
+    �621dee� + �04� + �01� + �000000000000000000000000000000000000000000000� +
```

Comment:
I think we should still hash the `location_id` to account for other location types. Simply concatenating a zero padded GLN to the prefix works well for GS1 because the GLN is a numeric string where each character is a number with a range of 0-9, but this addressing scheme won't work if we want to support a different location type that has a location id that contains characters that don't directly translate to hexadecimal.

As such, I would propose we change the address definition to something like

```
“621dee” + “04” + “01” + “hash(location_id)[:60]
```

Reply:
We should not have hashing in the address here. The initial construction was correct, and it is important that we not define an approach here that is inconsistent with the approach taken in product.

We hash when necessary, but it opens up the possibly of hash collisions. Here, we don't have to handle hash collisions. Other location namespaces can choose different techniques, so that argument is not valid -- "621dee0402" namespace for example may use hashing if necessary -- this RFC is not taking a stance on that.

Reply:
The implementation of the product rfc computes the address by [hashing the product id](https://github.com/hyperledger/grid/blob/e6e48ac37b7bb1563ab676c597025aac23b6a1da/contracts/product/src/addressing.rs#L38)
```
“621dee” + “02” + “01” + “hash(product_id)[:60]
```
So in order to maintain consistency between the two contracts we should either implement the addressing for location in a similar way (using hashing), or we should update product to not use hashing when calculating GS1 addresses.

Reply:
The rationale behind hashing the `product_id` to compute the address space assumes that `product_id` (which is a GTIN) is unique to the product, **_and_** only contains attribute specific to the product itself (i.e. dimensions, weight, etc). I believe this is described as "pre-competitive" data. 

It would not contain business-specific attributes such as price. Those business-specific attributes would be found in a `catalog_product`, which derives from a Grid `product`.  I believed this is described as competitive data.

To use this same approach for location, it'd need to be assumed that GLN's are unique and that location will only house "pre-competitive" data ([defined here](https://github.com/hyperledger/grid-rfcs/pull/20/files#diff-dff71a79f1740d09c3346aff673e2023R359)). If a location has business-specific attributes associated with it, there may need to be a subclass of the location to house that data.

This subclass implementation would follow the same pattern used between `product` and `catalog_product` to appropriately house pre-competitive/competitive data respectively. However, it seems that this "subclass of location", that would house competitive data, is out of scope for this RFC. 

I think it would be worth acknowledging in the drawback section of this RFC. Something like "Competitive data will not be captured in location but can be extended to at a later time if need be. Similarly to how catalog_product extends product."